pkg/zval/zval.go

```diff
@@ -15,6 +15,132 @@ import (
 	"fmt"
 )
 
+const (
+	beginContainer = -1
+	endContainer   = -2
+)
+
+type node struct {
+	innerLen int
+	outerLen int
+	dfs      int
+}
+
+// Builder implements an API for holding an intermediate representation
+// of a hierarchical set of values arranged in a tree, e.g., structured
+// values that can contain nested and recursive aggregate values.
+// We encode a DFS traversal in a flat data structure that can be
+// reused across invocations so we don't otherwise allocate a tree
+// data structure for every record parsed that would then be GC'd.
+type Builder struct {
+	nodes  []node
+	leaves [][]byte
+}
+
+func NewBuilder() *Builder {
+	return &Builder{
+		nodes:  make([]node, 0, 64),
+		leaves: make([][]byte, 0, 64),
+	}
+}
+
+func (b *Builder) Reset() {
+	b.nodes = b.nodes[:0]
+	b.leaves = b.leaves[:0]
+}
+
+func (b *Builder) Begin() {
+	b.nodes = append(b.nodes, node{dfs: beginContainer})
```

Comment:
Instead of allocating a new node shouldn't you be selecting/reseting a node from the slice's unused capacity?

Reply:
(This is kind of why I'd like to see a benchmark of this)

Reply:
It's not allocating a new node... just putting the struct into the pre-allocated slice.  It's the go way.

Reply:
I did know this!